raft/raft.cpp

```diff
@@ -792,49 +884,229 @@ raft::load_state()
+                std::set<bzn::uuid_t> result;
+                std::transform(
+                        peers.begin()
+                        , peers.end()
+                        , std::inserter(result, result.begin()),
+                        extract_uuid);
+                return std::list<std::set<bzn::uuid_t>>{result};
+            }
+            break;
+        case bzn::log_entry_type::joint_quorum:
+            {
+                std::set<bzn::uuid_t> result_old, result_new;
+                const auto& old_jpeers = log_entry.msg["msg"]["peers"]["old"];
+                const auto& new_jpeers = log_entry.msg["msg"]["peers"]["new"];
+                std::transform(old_jpeers.begin()
+                        , old_jpeers.end()
+                        , std::inserter(result_old, result_old.begin())
+                        , extract_uuid);
+            std::transform(new_jpeers.begin()
+                    , new_jpeers.end()
+                    , std::inserter(result_new, result_new.begin())
+                    , extract_uuid);
+            return std::list<std::set<bzn::uuid_t>>{result_old, result_new};
+            }
+            break;
+        default:
+            throw std::runtime_error("last_quorum gave something that's not a quorum");
+    }
+    return std::list<std::set<bzn::uuid_t>>();
+}
+
+
+bool
+raft::is_majority(const std::set<bzn::uuid_t>& votes)
+{
+    for(const auto& s : this->get_active_quorum())
+    {
+        std::set<bzn::uuid_t> intersect;
+        std::set_intersection(s.begin(), s.end(), votes.begin(), votes.end(),
+                              std::inserter(intersect, intersect.begin()));
+
+        if(intersect.size() * 2 <= s.size())
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+
+bzn::peers_list_t
+raft::get_all_peers()
+{
+    bzn::peers_list_t result;
+    std::vector<std::reference_wrapper<bzn::message>> all_peers;
+    bzn::log_entry log_entry = this->raft_log->last_quorum_entry();
 
-    if (this->log_entries.empty())
+    if(log_entry.entry_type == bzn::log_entry_type::single_quorum)
     {
-        throw std::runtime_error(MSG_ERROR_EMPTY_LOG_ENTRY_FILE);
+        all_peers.emplace_back(log_entry.msg["msg"]["peers"]);
     }
+    else
+    {
+        all_peers.emplace_back(log_entry.msg["msg"]["peers"]["old"]);
+        all_peers.emplace_back(log_entry.msg["msg"]["peers"]["new"]);
+    }
+
+    for(auto jpeers : all_peers)
+    {
+        for(const bzn::message& p : jpeers.get())
+        {
+            result.emplace(p["host"].asString(),
+                           uint16_t(p["port"].asUInt()),
+                           uint16_t(p["http_port"].asUInt()),
+                           p["name"].asString(),
+                           p["uuid"].asString());
+        }
+    }
+    return result;
+}
+
+
+bool
+raft::in_quorum(const bzn::uuid_t& uuid)
+{
+    const auto quorums = this->get_active_quorum();
+    for(const auto q : quorums)
```

Comment:
const auto& q?

Reply:
or:
for (const auto& q : this->get_active_quarun())
...

Reply:
Cleaned up.